### PR TITLE
Add SAI API docs

### PR DESCRIPTION
**Pull request by kegilbert**
@AnotherButler  @ithinuel 

First pass at SAI API docs. **Note** The example and API reference are not up to date. There may be a few segments that I'll highlight inline below that may be better for the contributing section.

### CODE REVIEW

File: docs/reference/api/drivers/SAI.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+SAI uses a four wire serial protocol consisting of:
+  * Master Clock Pin  (mclk)
+  * Data Pin          (sd)
+  * Bit Clock pin     (bclk)
+  * Word Clock Pin    (wclk)
+
+The master clock is used to optionally synchronize the device's IO lines. When using SAI to communicate in loopback or with an on-target audio codec there may be an internally shared master clock and this pin is not required. The `mclk_source` field in the format object that may be optionally passed into the SAI constructor can be set to internal, external or sibling.
+
+The bit clock fires on each bit of data transmitted across the data lines and its frequency is determined by the sampling rate, number of bits per channel, and number of channels. The sample rate is highly device dependent and needs to be configured for each target (typically found in the target's device.h file).
```
Comment (kegilbert) [inline docs/reference/api/drivers/SAI.md]: A good amount of this section may be better suited for the contributing docs page (pending). Thoughts?

File: docs/reference/api/drivers/SAI.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+ * `sai_init()` returns `SAI_RESULT_INVALID_PARAM`  if at least one of the given parameters is undefined (NULL)
+ * `sai_init()` returns `SAI_RESULT_ALREADY_INITIALIZED` if SAI is already initialized
+ * `sai_init()` returns `SAI_RESULT_CONFIG_UNSUPPORTED` if the device can never support this configuration
+ * `sai_init()` returns `SAI_RESULT_CONFIG_MISMATCH` if the device is not able to support this configuration at this point time because of other 'live' constraints
+ (such as a shared format/clock configuration with a sibling)
+ * `sai_free()` does nothing if passed a NULL pointer
+ * `sai_free()` de-initialized & un-clock unused part of the device a device/block can be reinitialized via `sai_init()` after being `sai_free()`d.
+
+### SAI class reference
```
Comment (kegilbert) [inline docs/reference/api/drivers/SAI.md]: Does not exist at the moment.

File: docs/reference/api/drivers/SAI.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+[![View code](https://www.mbed.com/embed/?type=library)](http://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/classmbed_1_1_sai.html)
+
+### SAI hello, world
+
+[![View code](https://www.mbed.com/embed/?url=https://os.mbed.com/teams/mbed_example/code/SAI_HelloWorld/)](https://os.mbed.com/teams/mbed_example/code/SAI_HelloWorld/file/fa13d56ff9ff/main.cpp)
```
Comment (kegilbert) [inline docs/reference/api/drivers/SAI.md]: Incorrect example title/links.

**Comment by AnotherButler** [comment]
@iriark01 @jen3andruska Could one of you please copy edit this?

**Comment by jen3andruska** [comment]
@AnotherButler No problem, looking at it now...

File: docs/reference/api/drivers/SAI.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  * Master Clock Pin  (mclk)
+  * Data Pin          (sd)
+  * Bit Clock pin     (bclk)
+  * Word Clock Pin    (wclk)
+
+The master clock is used to optionally synchronize the device's IO lines. When using SAI to communicate in loopback or with an on-target audio codec, there may be an internally shared master clock and this pin is not required. The `mclk_source` field in the format object that may be optionally passed into the SAI constructor can be set to internal, external or sibling.
+
+The bit clock fires on each bit of data transmitted across the data lines, and its frequency is determined by the sampling rate, number of bits per channel and number of channels. The sampling rate is highly device-dependent and needs to be configured for each target (typically found in the target's device.h file).
+
+The word clock runs at the frequency as the sample rate. It is used to select which active audio channel is being transmitted.
```
Comment (jen3andruska) [inline docs/reference/api/drivers/SAI.md]: Word missing here: "The word clock runs at the [   ] frequency as the sample rate.  I'm guessing 'same' but  let me know?

File: docs/reference/api/drivers/SAI.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+The Serial Audio Interface (SAI) provides a class to communicate with audio devices. SAI is a broader feature that encompasses audio interfaces such as I2S.
+
+SAI uses a four wire serial protocol consisting of:
+  * Master Clock Pin  (mclk)
+  * Data Pin          (sd)
+  * Bit Clock pin     (bclk)
+  * Word Clock Pin    (wclk)
+
+The master clock is used to optionally synchronize the device's IO lines. When using SAI to communicate in loopback or with an on-target audio codec, there may be an internally shared master clock and this pin is not required. The `mclk_source` field in the format object that may be optionally passed into the SAI constructor can be set to internal, external or sibling.
```
Comment (jen3andruska) [inline docs/reference/api/drivers/SAI.md]: This section starts off well explaining the Master Clock, but the next sentence about the pin, is confusing - which pin? The Master Clock pin or the Data pin?  If the former, then it doesn't explain what the Data pin is used for at all.  This section explains the Master Clock, Bit Clock and Word Clock but should also explain what the pins themselves are used for or how the SAI uses them, and it needs to say something about the Data Pin.

File: docs/reference/api/drivers/SAI.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
+## Serial Audio Interface (SAI)
+
+The Serial Audio Interface (SAI) provides a class to communicate with audio devices. SAI is a broader feature that encompasses audio interfaces such as I2S.
+
+SAI uses a four wire serial protocol consisting of:
+  * Master Clock Pin  (mclk)
+  * Data Pin          (sd)
+  * Bit Clock pin     (bclk)
+  * Word Clock Pin    (wclk)
+
```
Comment (jen3andruska) [inline docs/reference/api/drivers/SAI.md]: "Pin" should be capitalized or lowercase consistently (the first two are capitalized, the next two are lowercase).  I'm changing all to lower case.

File: docs/reference/api/drivers/SAI.md (hunk 5 above)
Comment (ithinuel) [reply docs/reference/api/drivers/SAI.md]: The Master Clock pin is not required to transfer data.

The data are sent on the _data pin_ using the _bit clock_ to identify when the data should be sampled.
the _word clock_ usually serves as word delimiter.

File: docs/reference/api/drivers/SAI.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+### Errors
+
+ * `sai_init()` returns `SAI_RESULT_INVALID_PARAM`  if at least one of the given parameters is undefined (NULL)
+ * `sai_init()` returns `SAI_RESULT_ALREADY_INITIALIZED` if SAI is already initialized
+ * `sai_init()` returns `SAI_RESULT_CONFIG_UNSUPPORTED` if the device can never support this configuration
+ * `sai_init()` returns `SAI_RESULT_CONFIG_MISMATCH` if the device is not able to support this configuration at this point in time because of other 'live' constraints
+ (such as a shared format/clock configuration with a sibling)
+ * `sai_free()` does nothing if passed a NULL pointer
+ * `sai_free()` de-initialized & un-clock unused part of the device a device/block can be reinitialized via `sai_init()` after being `sai_free()`d.
```
Comment (jen3andruska) [inline docs/reference/api/drivers/SAI.md]: This last description is confusing.  Do you mean that `sai_free()` [is] de-initialized?  If so, say: "[is] de-initialized & the un-clock [is] unused," but this seems like it should be followed by an explanation "if...." like in the previous description ("does nothing [if] passed a NULL pointer).   I can't figure out what 'part of the device' is doing?  Is the un-clock an unused part of the device?  When is it de-initialized & the 'un-clock unused' (this also sounds like a double negative - do you mean 'un-clock'?  and what is that?) . Need clarification, and also needs divisions.  "A device/block can be reinitialized via..." should start a new sentence.

File: docs/reference/api/drivers/SAI.md (hunk 5 above)
Comment (jen3andruska) [reply docs/reference/api/drivers/SAI.md]: This needs to be clarified better.

File: docs/reference/api/drivers/SAI.md (hunk 7 above)
Comment (ithinuel) [reply docs/reference/api/drivers/SAI.md]: This should actually read
```
`sai_free()` de-initializes & un-clocks unused part of the device. A device/block can be reinitialized via `sai_init()` after being `sai_free()`d.
```

Comment (jen3andruska) [reply docs/reference/api/drivers/SAI.md]: Ok, I see. This makes more sense now.

File: docs/reference/api/drivers/SAI.md (hunk 4 above)
Comment (kegilbert) [reply docs/reference/api/drivers/SAI.md]: Correct, 'same' is the right word there thanks!

**Comment by AnotherButler** [comment]
Note: This is not on master yet. We'll merge once the code does.

**Comment by ithinuel** [comment]
In order to keep a "clean history" you might also want to squash all the "Update SAI.md" as a single commit.

**Comment by AnotherButler** [comment]
I'm closing this until this feature is planned. Please feel to reopen then.

**Pull request closed without merging**